firestore/src/jni/arena_ref.cc

```diff
@@ -0,0 +1,147 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "firestore/src/jni/arena_ref.h"
+
+#include <atomic>
+#include <utility>
+
+#include "firestore/src/android/firestore_android.h"
+#include "firestore/src/jni/env.h"
+#include "firestore/src/jni/hash_map.h"
+#include "firestore/src/jni/long.h"
+#include "firestore/src/jni/map.h"
+
+namespace firebase {
+namespace firestore {
+namespace jni {
+
+namespace {
+
+HashMap* gArenaRefHashMap = nullptr;
+jclass gHashMapClass = nullptr;
+jclass gLongClass = nullptr;
+jmethodID gLongConstructor = nullptr;
+std::mutex mutex_;
+
+int64_t GetNextArenaRefKey() {
+  static std::atomic<int64_t> next_key(0);
+  return next_key.fetch_add(1);
+}
+
+}  // namespace
+
+ArenaRef::ArenaRef(Env& env, const Object& object)
+    : key_(GetNextArenaRefKey()) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  gArenaRefHashMap->Put(env, key_object(env), object);
+}
+
+ArenaRef::ArenaRef(const ArenaRef& other)
+    : key_(other.key_ == -1 ? -1 : GetNextArenaRefKey()) {
+  if (other.key_ != -1) {
+    Env env = FirestoreInternal::GetEnv();
+    Local<Object> object = other.get(env);
+    std::unique_lock<std::mutex> lock(mutex_);
+    gArenaRefHashMap->Put(env, key_object(env), object);
+  }
+}
+
+ArenaRef::ArenaRef(ArenaRef&& other) { std::swap(key_, other.key_); }
+
+ArenaRef& ArenaRef::operator=(const ArenaRef& other) {
+  Env env = FirestoreInternal::GetEnv();
+
+  if (this == &other) {
+    return *this;
+  }
+
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (key_ != -1) {
+      gArenaRefHashMap->Remove(env, key_object(env));
+      key_ = -1;
+    }
+
+    if (other.key_ != -1) {
+      key_ = GetNextArenaRefKey();
+      Local<Object> object = other.get(env);
+      gArenaRefHashMap->Put(env, key_object(env), object);
+    }
+  }
+  return *this;
+}
+
+ArenaRef& ArenaRef::operator=(ArenaRef&& other) {
+  Env env = FirestoreInternal::GetEnv();
+
+  if (this == &other) {
+    return *this;
+  }
+
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (key_ != -1) {
+      gArenaRefHashMap->Remove(env, key_object(env));
+    }
+    key_ = other.key_;
+    other.key_ = -1;
+  }
+  return *this;
+}
+
+ArenaRef::~ArenaRef() {
+  if (key_ != -1) {
+    Env env;
+    ExceptionClearGuard block(env);
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      gArenaRefHashMap->Remove(env, key_object(env));
+    }
+    if (!env.ok()) {
+      env.ExceptionDescribe();
+      env.ExceptionClear();
+    }
+  }
+}
+
+Local<Long> ArenaRef::key_object(Env& env) const {
+  jobject key = env.get()->NewObject(gLongClass, gLongConstructor, key_);
```

Comment:
If we add a unit test to `key_object()`, then this function needs to be public, since the test class is inside anonymous namespace. To me `key_object ` is more like an internal implementation, as long as we test out the behaviour of `ArenaRef` is correct we don't need to worry about how it is implemented.